vpx-encoder: doEncode to use forceKeyFrame

chivakker opened the pull request:
doEncode has to use forceKeyFrame to request a key frame to the driver. There's
no need for the libyami component to calculate the next key frame itself.

Application will set the key frame flag on its own.

Signed-off-by: Daniel Charles daniel.charles@intel.com

Fixes #583

xuguangxin commented on encoder/vaapiencoder_vp9.cpp:
```diff
@@ -159,9 +157,7 @@ YamiStatus VaapiEncoderVP9::doEncode(const SurfacePtr& surface,
-    m_frameCount %= keyFramePeriod();
-    picture->m_type = (m_frameCount ? VAAPI_PICTURE_P : VAAPI_PICTURE_I);
-    m_frameCount++;
+    picture->m_type = (forceKeyFrame ? VAAPI_PICTURE_I : VAAPI_PICTURE_P);
```
Force Key frame usually use by realtime communication program. If there are some frame drop between peers, following frames are not decodeable, until peer see a key frame. So If peer detect a frame dropping, It will sent a notification back to yami’s user,  then the user will sent an forceKeyFrame to yami. Therefore, we’d better have two machnism here.
1.  The normal case, the I frame decide by intraPeriod
2.  If user want force key frame, we will start a new gop.
Please refer to https://github.com/01org/libyami/blob/apache/encoder/vaapiencoder_h264.cpp#L1154
And https://github.com/01org/libyami/blob/apache/encoder/vaapiencoder_h264.cpp#L1279

chivakker replied:
In the case of VPX there's no need to reset the GOP.  But it is correct that the intraPeriod has to be handled separately from the forceKeyFrame.  The example provided on libyami-utils will force a keyFrame based on intraPeriod but it has to consider the other scenarios and involve other parameters like kf_min_dist kf_max_dist and kf_auto.  I'll submit a better example for this based also on how the driver is supposed to interpret the kf_\* variables.

xuguangxin replied:
Reset gop or not will depends on the implementation. However, here are some think about this:
Suppose key period is 30, User request key frame at 29. If we do not reset the GOP, it will generate a key frame in 29 and 30, this is unnecessary and will make bitrate very high, at this point.
Base on this we choose to reset GOP at h264 encoder

xuguangxin reviewed (changes requested):
You may have misunderstand about forceKeyFrame;)

chivakker committed:
vpx-encoder: doEncode to set key frame with  forceKeyFrame

doEncode has to use forceKeyFrame to request a key frame to the driver.

Application will set the key frame flag on its own

vpxencoder m_frameCount variable will now count all frames processed and
requesting a keyFrame at the right intraPeriod request from the application

Signed-off-by: Daniel Charles <daniel.charles@intel.com>

chivakker commented:
Re-submitted the patch considering the comments form @xuguangxin.  There's still some room for improvement once kf_\* variables are better understood.  There's no need to change anything on libyami-utils.

xuguangxin commented:
lgtm, thanks

xuguangxin commented:
merged as 0e9c1051495ff460eb5f5946940a682948f3d5cd
